src/main/java/org/openstreetmap/atlas/checks/validation/areas/WaterbodyAndIslandSizeCheck.java

```diff
@@ -0,0 +1,310 @@
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        if (object instanceof Area)
+        {
+            final Area area = (Area) object;
+            final Surface surfaceArea = area.asPolygon().surface();
+            final double surfaceAreaMeters = surfaceArea.asMeterSquared();
+            final double surfaceAreaKilometers = surfaceArea.asKilometerSquared();
+            // Mark each object because we could potentially run into Multipolygon entities that get
+            // passed in as Areas
+            this.markAsFlagged(area.getIdentifier());
+
+            // natural=water tag
+            if (TagPredicates.IS_WATER_BODY.test(object))
+            {
+                if (surfaceAreaMeters < this.waterbodyMinimumArea)
+                {
+                    return Optional.of(this.createFlag(object,
+                            this.getLocalizedInstruction(0, object.getOsmIdentifier(),
+                                    surfaceAreaMeters, this.waterbodyMinimumArea)));
+                }
+                else if (surfaceAreaKilometers > this.waterbodyMaximumArea)
+                {
+                    return Optional.of(this.createFlag(object,
+                            this.getLocalizedInstruction(1, object.getOsmIdentifier(),
+                                    surfaceAreaKilometers, this.waterbodyMaximumArea)));
+                }
+            }
+            // place=island or place=archipelago tags
+            if (IS_ISLAND.test(object))
+            {
+                if (surfaceAreaMeters < this.islandMinimumArea)
+                {
+                    return Optional.of(this.createFlag(object, this.getLocalizedInstruction(2,
+                            object.getOsmIdentifier(), surfaceAreaMeters, this.islandMinimumArea)));
+                }
+                else if (surfaceAreaKilometers > this.islandMaximumArea)
+                {
+                    return Optional.of(this.createFlag(object,
+                            this.getLocalizedInstruction(THREE, object.getOsmIdentifier(),
+                                    surfaceAreaKilometers, this.islandMaximumArea)));
+                }
+            }
+            // place=islet tag
+            if (IS_ISLET.test(object))
+            {
+                if (surfaceAreaMeters < this.isletMinimumArea)
+                {
+                    return Optional.of(this.createFlag(object, this.getLocalizedInstruction(FOUR,
+                            object.getOsmIdentifier(), surfaceAreaMeters, this.isletMinimumArea)));
+                }
+                else if (surfaceAreaKilometers > this.isletMaximumArea)
+                {
+                    return Optional.of(this.createFlag(object,
+                            this.getLocalizedInstruction(FIVE, object.getOsmIdentifier(),
+                                    surfaceAreaKilometers, this.isletMaximumArea)));
+                }
```

Comment:
As mentioned before instructions for min/max could be combined, which would make this part shorter and cleaner.